File: software/examples/rust/example_threshold.rs
use std::{error::Error, io, thread};
use tinkerforge::{ambient_light_v3_bricklet::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Ambient Light Bricklet 3.0.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let al = AmbientLightV3Bricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    let illuminance_receiver = al.get_illuminance_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `al` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for illuminance in illuminance_receiver {
            println!("Illuminance: {} lx", illuminance as f32 / 100.0);
            println!("Too bright, close the curtains!");
        }
    });

    // Configure threshold for illuminance "greater than 500 lx"
    // with a debounce period of 1s (1000ms).
    al.set_illuminance_callback_configuration(1000, false, '>', 500 * 100, 0);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/csharp/ExampleThreshold.cs
using System;
using Tinkerforge;

class Example
{
	private static string HOST = "localhost";
	private static int PORT = 4223;
	private static string UID = "XYZ"; // Change XYZ to the UID of your Ambient Light Bricklet 3.0

	// Callback function for illuminance callback
	static void IlluminanceCB(BrickletAmbientLightV3 sender, long illuminance)
	{
		Console.WriteLine("Illuminance: " + illuminance/100.0 + " lx");
		Console.WriteLine("Too bright, close the curtains!");
	}

	static void Main()
	{
		IPConnection ipcon = new IPConnection(); // Create IP connection
		BrickletAmbientLightV3 al = new BrickletAmbientLightV3(UID, ipcon); // Create device object

		ipcon.Connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Register illuminance callback to function IlluminanceCB
		al.IlluminanceCallback += IlluminanceCB;

		// Configure threshold for illuminance "greater than 500 lx"
		// with a debounce period of 1s (1000ms)
		al.SetIlluminanceCallbackConfiguration(1000, false, '>', 500*100, 0);

		Console.WriteLine("Press enter to exit");
		Console.ReadLine();
		ipcon.Disconnect();
	}
}
